examples/EF_QAE/main.py

```diff
@@ -0,0 +1,265 @@
+    
+    def ansatz_EF_QAE(theta, x):
+        """Creates the variational quantum circuit for EF-QAE.
+        Args:
+            theta (array or list): values of the parameters.
+            x (float): value of the input feature
+
+        Returns:
+            Quantum circuit.
+        """
+        circuit = models.Circuit(nqubits)
+        index = 0
+        for l in range(layers):
+            for q in range(nqubits):
+                circuit.add(gates.RY(q, theta[index]*x + theta[index+1]))
+                index+=2
+            circuit.add(gates.CZ(5, 4))
+            circuit.add(gates.CZ(5, 3))
+            circuit.add(gates.CZ(5, 1))
+            circuit.add(gates.CZ(4, 2))
+            circuit.add(gates.CZ(4, 0))
+            for q in range(nqubits):
+                circuit.add(gates.RY(q, theta[index]*x + theta[index+1]))
+                index+=2
+            circuit.add(gates.CZ(5, 4))
+            circuit.add(gates.CZ(5, 2))
+            circuit.add(gates.CZ(4, 3))
+            circuit.add(gates.CZ(5, 0))
+            circuit.add(gates.CZ(4, 1))
+        for q in range(nqubits-compress, nqubits, 1):
+            circuit.add(gates.RY(q, theta[index]*x + theta[index+1]))
+            index+=2
+        return circuit
+
+    cost_function_steps = []
+    nqubits = 6
+    compress = 2
+    encoder = encoder_hamiltonian_simple(nqubits, compress)
+    count = [0]
+    
+    if example == 0:
+        ising_groundstates = []
+        lambdas = np.linspace(0.5, 1.0, 20)
+        for lamb in lambdas:
+            ising_ham = -1 * hamiltonians.TFIM(nqubits, h=lamb)
+            ising_groundstates.append(ising_ham.eigenvectors()[0])
+            
+        if autoencoder == 1:
+            def cost_function_QAE_Ising(params, count):
+                """Evaluates the cost function to be minimized for the QAE and Ising model.
+        
+                Args:
+                    params (array or list): values of the parameters.
+        
+                Returns:
+                    Value of the cost function.
+                """
+                circuit = models.Circuit(nqubits)
+                for l in range(layers):
+                    for q in range(nqubits):
+                        circuit.add(gates.RY(q, theta=0))
+                    circuit.add(gates.CZ(5, 4))
+                    circuit.add(gates.CZ(5, 3))
+                    circuit.add(gates.CZ(5, 1))
+                    circuit.add(gates.CZ(4, 2))
+                    circuit.add(gates.CZ(4, 0))
+                    for q in range(nqubits):
+                        circuit.add(gates.RY(q, theta=0))
+                    circuit.add(gates.CZ(5, 4))
+                    circuit.add(gates.CZ(5, 2))
+                    circuit.add(gates.CZ(4, 3))
+                    circuit.add(gates.CZ(5, 0))
+                    circuit.add(gates.CZ(4, 1))
+                for q in range(nqubits-compress, nqubits, 1):
+                    circuit.add(gates.RY(q, theta=0))
```

Comment:
Here the circuit is defined inside the cost function, so using `set_parameters` won't really make any difference. The circuit will again be recreated every time the cost function is called during optimization, so it is essentially the same without using `set_parameters`. 
If transforming this to the `set_parameters` format is complicated and this should be merged asap then we could leave it as it was before and postpone the update for later.

Reply:
I believe the last updated should be fine regarding to that, I manage to do it with Stefano's help.

Reply:
Thanks for the update, it seems to be working well now. The only thing I think we need to change is to move the four circuit definitions outside of the cost functions. This way we will avoid recreating the circuit again and again during optimization. You can just move it on top, outside of each function, eg:
```Python
circuit = models.Circuit(nqubits)
for l in range(layers):
    for q in range(nqubits):
        circuit.add(gates.RY(q, theta=0))
     circuit.add(gates.CZ(5, 4))
     ...

def cost_function_EF_QAE_Digits(params, count):
    """..."""
    cost = 0
    newparams = rotate(params, 1)
    circuit.set_parameters(newparams)
    ...
```
I tested this and it seems to work as expected (in terms of the printed losses).

Regarding the test that fails, I think these segfaults are due to the time-out so there is probably nothing you can do. We need to look at this.

Reply:
Done. Thanks Stavros.

Is not possible then to approve the PR?